python: add `QueueList` class for access to queue config, limits, status and resource counts

grondo opened the pull request:
Problem: There's currently no easy way to access Flux queue information from the Python API, and that information is currently distributed between different parts the config, various job-manager RPCs, and resource configuration.

This PR adds a new `flux.queue.QueueList` class which fetches information for all currently configured queues in one go. It then presents as a list of `QueueInfo` objects for each queue, which offer access to status, limits, and even resources counts on a per queue basis. This class is then used to replace similar code in `flux queue list`.

In a follow on PR, `flux queue list` will be extended to allow access to the resource stats.

jameshcorbett commented on t/python/t0034-queuelist.py:
```diff
@@ -0,0 +1,100 @@
+        """
+        self.fh.rpc("config.load", tomllib.loads(testconf)).get()
+
+        qlist = QueueList(self.fh)
```
Maybe add a test like `QueueList(self.fh, [])`?

chu11 replied:
and also requesting specific queues, `QueueList(self.fh, ["batch"]`, and probably multiple queues `["batch", "debug"]`

jameshcorbett reviewed (approved):
A bunch of small comments but generally looks great!

chu11 commented on src/bindings/python/flux/queue.py:
```diff
@@ -0,0 +1,345 @@
+attributes are also documented for each corresponding class in this module::
+
+  queue.name (str) (empty string if using a single anonymous queue)
+  queue.is_default (str)
```
is_default -> bool?

grondo commented on src/bindings/python/flux/resource/list.py:
```diff
@@ -82,6 +84,7 @@ def free(self):
         return res
 
 
+
```
Hm, I'm assuming `black` added this, but I can try removing it and see if it complains.

grondo replied:
Oh I get it, as @chu11 said, in the wrong commit.

chu11 commented on src/bindings/python/flux/queue.py:
```diff
@@ -0,0 +1,345 @@
+
+    min: QueueResourceCounts
+    max: QueueResourceCounts
+    duration: str
```
str -> float?

grondo commented on src/cmd/flux-queue.py:
```diff
@@ -327,17 +327,13 @@ def list(args):
-            config = future.get()
-        except Exception as e:
-            LOGGER.warning("Could not get flux config: " + str(e))
+    future = handle.rpc("config.get")
```
Great suggestion! However, this particular commit was updating existing code and I'd prefer not to make two changes at once. Actually, I see there are several uses of `handle.rpc("config.get")` throughout the code that should all be updated, let's do that in a separate PR.

grondo commented on src/bindings/python/flux/queue.py:
```diff
@@ -0,0 +1,345 @@
+    Attributes:
+        min (:obj:`QueueResourceCounts`): Configured resource minimums.
+        max (:obj:`QueueResourceCounts`): Configured resource maximums.
+            (If no maximum, individual resource count will be ``inf``)
```
Thanks! I'm not all that familiar with Python typing hints. Is there a more generic "Number" type hint?

grondo replied:
Ah from [PEP 484](https://peps.python.org/pep-0484/#the-numeric-tower)

> when an argument is annotated as having type float, an argument of type int is acceptable; 

I'll use `float`

grondo commented on src/bindings/python/flux/queue.py:
```diff
@@ -0,0 +1,345 @@
+    """
+
+    def __init__(self, handle, queues=None):
+        resources, config = map(
+            lambda x: x.get(),
+            [resource_list(handle), handle.rpc("config.get")],
+        )
```
This should have a comment perhaps, but we don't want to run the RPCs back-to-back. (This code initiates both RPCs before calling get on either (at least I hope that's what it does)). May need to change anyway if we update to `handle.conf_get()` - though now I'm realizing there's no way to make that call asynchronous, which is why the explicit RPC may be better here :thinking:

chu11 commented on src/cmd/flux-queue.py:
```diff
@@ -144,161 +143,78 @@ def validate(self, path, conf):
+        return getattr(self.__limits, attr)
 
 
 class QueueInfo:
```
to avoid confusion with the new QueueInfo class, rename to  ....???

jameshcorbett commented on src/bindings/python/flux/queue.py:
```diff
@@ -0,0 +1,345 @@
+ >>> batchq = qlist["batch"]
+
+If there is a single anonymous queue, the queue name is the empty string,
+so this queue may be indexed using an empty string as the key:
```
Ah ok makes sense. That could be a change in the `flux queue list` command though right? Up to you though.

grondo commented on src/bindings/python/flux/queue.py:
```diff
@@ -0,0 +1,345 @@
+        If ``queues`` is None, return the whole config, which may be empty
+        if there is no configured queues.
+        """
+        if not queues:
```
This is done purposely to easily support callers that are passing argparse `args.queue` option directly which is of type append or Flux's `FilterSetUpdate`, where the default (no queue specifically specified) value is an empty list or set.

I will update the documentation accordingly.

grondo commented:
Thanks for all the good feedback @jameshcorbett  and @chu11! Very helpful.

I've addressed most of your comments and forced a push. For comments I did not address I gave a reason above.

jameshcorbett commented:
Looks great!

chu11 reviewed (approved):
LGTM!

grondo commented:
Thanks! Seeing MWP
